### ISSUE TITLE
`DiagnosticSource` Filtering Behaviour

### ISSUE THREAD

**Issue by stevejgordon** [dotnet/runtime#98499]
While digging into the implementation details and advanced scenarios for `DiagnosticSource`, I ran into something slightly surprising that was not covered by the docs. If you subscribe multiple observers to the same notification source, those subscribing with a filter may still receive notifications from `OnNext` if any other subscribers require the notification.

**Repro:**
```c#
using System.Diagnostics;
using Test;

DiagnosticListener.AllListeners.Subscribe(new DiagnosticListenersObserver());

DoStuff.DoIt();

namespace Test
{
    public class DoStuff
    {
        internal static readonly DiagnosticListener Listener = new("MySource");

        public static void DoIt()
        {
            if (Listener.IsEnabled("DoIt"))
                Listener.Write("DoIt", null);
        }
    }

    public class DiagnosticListenersObserver : IObserver<DiagnosticListener>
    {
        public void OnCompleted() { }
        public void OnError(Exception error) { }

        public void OnNext(DiagnosticListener value)
        {
            if (value.Name == "MySource")
            {
                value.Subscribe(new Observer("Unfiltered"));
                value.Subscribe(new Observer("Filter everything"), n => false);
            }
            else
                Console.WriteLine(value.Name);
        }
    }

    public class Observer(string name) : IObserver<KeyValuePair<string, object?>>
    {
        public string Name { get; } = name;

        public void OnCompleted() { }
        public void OnError(Exception error) { }

        public void OnNext(KeyValuePair<string, object?> value) => 
            Console.WriteLine($"{Name}: {value.Key}");
    }
}
```

**Console output:**
```
Filter everything: DoIt
Unfiltered: DoIt
```

As far as I can see, this is not explicitly called out in the docs, so I wanted to validate this is expected. Presumably, observers subscribed with a filter must recheck their filtering to handle only intended notifications in `IObserver<KeyValuePair<string, object?>>.OnNext`. If these assumptions are correct, I can raise a docs issue for this and perhaps help with a PR to add a note to the docs. 

On a related point, I have a query about the guidance in https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md

Under the payloads section, it states

> Note that this type needs to be made public (since the listener needs to see it) and should be under the namespace System.Diagnostics.DiagnosticSource.PayloadTypes.

If this advice is followed, however, when accessing the payload type in an observer, we would require a using directive to that namespace, which causes compiler warning CS0435, due to the namespace conflicting with the `DiagnosticSource` type. Do you think this recommendation should be amended or removed?

**Comment by tommcdon** [comment]
@noahfalk

**Comment by noahfalk** [comment]
Hi @stevejgordon, thanks for taking the time to make such a nice write up!

> As far as I can see, this is not explicitly called out in the docs, so I wanted to validate this is expected

As best I can tell from unit tests for this feature the filtering behavior you are seeing is by design:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/tests/DiagnosticSourceTests.cs#L242-L256

As the tests describe it the filters are intended as a performance optimization but not a guarantee. If the source of the data doesn't call IsEnabled() prior to calling Write() that would be another case where a listener might get events despite having a filter that would have rejected it.

> If these assumptions are correct, I can raise a docs issue for this and perhaps help with a PR to add a note to the docs.

I created https://github.com/dotnet/runtime/pull/98801 to address this, but if you had a different change in mind an issue/PR is very welcome :)

> Do you think this recommendation should be amended or removed?

I think we should remove that naming guidance.
- From a little cursory searching I can't find a single example of this guidance being followed.
- It conflicts with our naming guidance in https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-namespaces.

**Comment by stevejgordon** [comment]
You're welcome @noahfalk.

Thanks for double-checking. I should have thought to review the test cases also! 

That change in #98801 looks good. It would probably be worth also porting the same note into the [learn docs page](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/diagnosticsource-diagnosticlistener), also. I'll be honest, that doc is quite hard to follow with the complex example it includes, but at least the filtering note would avoid surprises.
